### DISCUSSION TITLE
Could we get a single-header-file version?

### DISCUSSION THREAD

**Discussion by Toddler-Boy** [fktn-k/fkYAML#255] [Ideas]
For nlohmann/json, there exists a version in which all files have been amalgamated into a single header file, which makes usage much easier. If we want to update to the lastest json parser, we just download the one file and we're done. Same for usage. You just include the one file where you need to parse json, and you're done.

I think it would be wonderful if fkYAML could also provide an amalgamated version. It could be auto-generated for every commit and/or release.

**Comment by fktn-k** [comment]
Thanks for sharing your idea.  
And I agree with your suggestion.  

Though I've been thinking about it too, I couldn't resolve several concerns, such as:  

* what amalgamation tool should be used considering the long term,
* whether we should update the amalgamated header per commit or per release, or
* in what way the amalgamation process should be provided so we can easily run it.

Do you have any suggestion?

**Comment by Toddler-Boy** [comment]
If PHP is permissible, I could create a simple tool that loads all header files in a particular order, removes all #include statements referring to these files, and outputs a single amalgamated file.

It's relatively simple. If PHP is not an option, then I could also write it in C++, it would just take longer.

I think it would be a good idea to run this for every commit, unless you plan to have releases often for bug fixes or new features.

**Reply by fktn-k** [reply]
Thanks for your suggestion.  
As for the frequency of amalgamation, I think it would be sufficient to force contributors to amalgamate headers before merging PRs, since people won'tcare about other branches than main/develop.  

Back to your suggestions.  
IMHO both solutions using PHP and C++ are not desirable. (I should have told you minimum requirements I was thinking for the amalgamation tool. Sorry about that.)  

One reason is that I don't want to add extra dependencies to this project as much as possible, and also don't want to  force contributors to install PHP only for amalgamation. I guess amalgamate.py will do just like nlohmann/json uses for amalgamation, but am I missing something?  

Another reason is that I just want to avoid buildings just for the tool written in C++ if it's not unavoidable.

**Reply by Toddler-Boy** [reply]
Then I won't be able to help, as I've never written a Python script. But I guess you could copy the one from nlohmann/json and adapt it to your needs.

**Reply by fktn-k** [reply]
You are helping! I'll work on that this weekend, hopefully.

**Reply by Toddler-Boy** [reply]
Please don't rush this on my behalf. It's not that urgent.

**Reply by fktn-k** [reply]
@reFX-Mike   
I've made changes to support the single-header-file version in the PR #261.  
Since all the workflows successfully passed with the single-header-file version, it should be fine to use it on your environment as well.  
Let me know if I missed something.  

Also, the single-header-version will be available from release pages from the next release on.
